### PR TITLE
Email report: identify test cases that fail due to involvement of two or more infras

### LINKED ISSUE #919: Identify the exact infrastructure a given test case has failed
**Description:**

Using statistical analysis of builds of the test plans in the current iteration, we should try to determine the exact failing infrastructure that lead to test failures/errors. This is much useful than showing lists of failing infrastructure combinations. 

image

**Affected Product Version:**
m36

**Related Issues:**
#906

### PR DESCRIPTION
**Pull request by kasunbg**
**Purpose**
Using statistical analysis of builds of the test plans in the current iteration, we should try to determine the exact failing infrastructure that lead to test failures/errors. This is much useful than showing lists of failing infrastructure combinations.

Fixes #919 

**Goals**
Better presentation of testgrid results.

### CODE REVIEW

File: common/src/main/java/org/wso2/testgrid/common/infrastructure/InfrastructureParameter.java [hunk 1]
```diff
@@ -188,4 +242,60 @@ public int hashCode() {
...
+            clone.readyForTestGrid = this.readyForTestGrid;
+            clone.properties = this.properties;
+            clone.subInfrastructureParameters = this.subInfrastructureParameters;
+            return clone;
+
+        } catch (CloneNotSupportedException e) {
+            throw new TestGridError("Since the super class of this object is java.lang.Object that supports cloning, "
+                    + "this failure condition should never happen unless a serious system error occurred.", e);
+        }
+
```
Comment (yasassri) [inline common/src/main/java/org/wso2/testgrid/common/infrastructure/InfrastructureParameter.java]: Unnecessary space.

File: common/src/main/java/org/wso2/testgrid/common/util/FileUtil.java [hunk 2]
```diff
@@ -234,12 +234,12 @@ private static void createFileIfNotExists(String filePath) throws TestGridExcept
         }
         try (Stream<Path> stream = Files.list(source).filter(Files::isRegularFile)) {
             List<Path> paths = stream.sorted().collect(Collectors.toList());
+            logger.info("A test plan files found at " + paths);
```
Comment (yasassri) [inline common/src/main/java/org/wso2/testgrid/common/util/FileUtil.java]: Testplan files found at?

File: common/src/main/java/org/wso2/testgrid/common/infrastructure/InfrastructureParameter.java [hunk 3]
```diff
@@ -125,6 +144,41 @@ public void setReadyForTestGrid(boolean readyForTestGrid) {
...
+                    infraParameter.setName(name);
+                    infraParameter.setType(type);
+                    infraParameter.setReadyForTestGrid(true);
+                    infraParameter.setProperties(this.getProperties());
+                    subInfrastructureParameters.add(infraParameter);
+                }
+            } catch (IOException e) {
+                logger.warn("Error while loading the infrastructure parameter's properties string for: " + this);
+            }
+
```
Comment (lasanthaDLPDS) [inline common/src/main/java/org/wso2/testgrid/common/infrastructure/InfrastructureParameter.java]: Shall we remove the unnecessary new line?

File: common/src/main/java/org/wso2/testgrid/common/util/TestGridUtil.java [hunk 4]
```diff
@@ -147,6 +147,38 @@ public static Path getTestRunWorkspace(TestPlan testPlan, boolean relative) {
...
+    public static Map<String, String> parseInfraParametersString(String infraParams) {
+        try {
+            ObjectMapper mapper = new ObjectMapper();
+            return mapper.readValue(infraParams, new TypeReference<Map<String, String>>() {
+            });
+        } catch (IOException e) {
+            logger.error("Error while parsing infra parameters", e);
+            return Collections.emptyMap();
+        }
+
```
Comment (lasanthaDLPDS) [inline common/src/main/java/org/wso2/testgrid/common/util/TestGridUtil.java]: Shall we remove the unnecessary new line?

File: common/src/main/java/org/wso2/testgrid/common/util/TestGridUtil.java [hunk 5]
```diff
@@ -147,6 +147,38 @@ public static Path getTestRunWorkspace(TestPlan testPlan, boolean relative) {
...
+     *
+     */
+    public static String convertToJsonString(Properties properties) {
+        try {
+            return new ObjectMapper().writeValueAsString(properties);
+        } catch (JsonProcessingException e) {
+            logger.error("Error while transforming to json string", e);
+            return null;
+        }
+
```
Comment (lasanthaDLPDS) [inline common/src/main/java/org/wso2/testgrid/common/util/TestGridUtil.java]: Shall we remove the unnecessary new line?

File: reporting/src/main/java/org/wso2/testgrid/reporting/EmailReportProcessor.java [hunk 6]
```diff
@@ -213,6 +224,42 @@ public boolean hasFailedTests(List<TestPlan> testPlans) {
      * @return summary table
      */
     public Map<String, InfrastructureBuildStatus> getSummaryTable(List<TestPlan> testPlans) {
-        return new InfrastructureSummaryReporter().getSummaryTable(testPlans);
+        return infrastructureSummaryReporter.getSummaryTable(testPlans);
+    }
+
+
```
Comment (lasanthaDLPDS) [inline reporting/src/main/java/org/wso2/testgrid/reporting/EmailReportProcessor.java]: Shall we remove unnecessary new lines?

File: reporting/src/test/java/org/wso2/testgrid/reporting/BaseClass.java [hunk 7]
```diff
@@ -227,43 +243,84 @@ protected Path getTestPlansPath() {
...
+            ts.addTestCase(tc);
+            tc.setTestScenario(ts);
+            ts.setStatus(tc.getStatus());
+        } else {
+            TestCase tc = createTestScenarioFor("CentOS-OPEN_JDK8-APIInvocationStatPublisherTestCase", Status.SUCCESS,
+                    "I passed.");
+            ts.addTestCase(tc);
+            tc.setTestScenario(ts);
+        }
+
```
Comment (lasanthaDLPDS) [inline reporting/src/test/java/org/wso2/testgrid/reporting/BaseClass.java]: Shall we remove unnecessary new lines?

**Pull request merged**